### DISCUSSION TITLE
ESP was frozen - reason MQTT queue jam?

### DISCUSSION THREAD

**Discussion by kalleick** [dewenni/ESP_Buderus_KM271#29] [Q&A]
Question:

My ESP was frozen after my MQTT broker was crashed, I don't know why. I had to turn it off and on.

Is there an MQTT queue where the values were collected that have not yet been sent?

I made this observation: my MQTT broker was unavailable for 30 minutes.
During this time the ESP32 was frozen. After the MQTT broker came back, the ESP32 still could not be operated.
Was that coincidence? Does a broker error mess up the ESP? If so, is there a way to stabilize the ESP?

**Comment by dewenni** [comment]
If mqtt is enabled, there is a cyclic function that check if the connection is valid.
If the connection to the broker is lost, the esp tries 5 times to reconnect.
Between every try there is a delay time of 10 seconds.
If the esp can not connect after 5 tires, the esp will restart and then try again.

Could it be that you have seen this behavior?

**Comment by kalleick** [comment]
thank you very much for the explanation.
An automatic ESP restart could not be recognized, it was several minutes stuck till i did a manual reboot. 
But I have not tested it intensively yet.

**Reply by dewenni** [reply]
Ok, I haven't seen this so far in my setup. Maybe it is something special on your side?!
If you have further infos, feel free to share it with me.
I am always open to help and improve.

**Reply by JieffeEl** [reply]
I just noticed the same problem. I rebooted my MQTT server (Mosquitto) after an OS update (so at most a few minutes downtime) and the ESP stopped responding. I once encountered that problem with another program (https://github.com/wdoekes/pe32me162ir_pub) but it wasn't even using the same MQTT library (that program used ArduinoMQTTClient, not PubSubClient as yours does).

**Reply by dewenni** [reply]
Ok, I will do some Tests in the next days and see if I could reproduce this also in my setup.
Up to now I haven't observed this.

**Reply by dewenni** [reply]
So, I did some tests today and I found out, that there is a problem with the "double reset detection".
Normally this should detect, if the reset button is pressed two times within 10 seconds.
But It seems, that this also triggered when the esp restarts two times because of a lost connection.
Then it activates the setup mode and enables the local access point.

@kalleick @JieffeEl could this be the behavior that you have recognized?


```
WiFi not connected. Trying connect to: xxx
WiFi reconnected
IP address: 
192.168.178.197
MQTT not connected, reconnect...
MQTT connected

... => manually stop mqtt broker

MQTT not connected, reconnect...
failed, rc=-4, retrying
MQTT not connected, reconnect...
failed, rc=-4, retrying
MQTT not connected, reconnect...
failed, rc=-4, retrying
MQTT not connected, reconnect...
failed, rc=-4, retrying
MQTT not connected, reconnect...
failed, rc=-4, retrying
MQTT connection not possible, rebooting...WS_EVT_DISCONNECT

... => reboot esp after 5 attempts

WiFi Mode STA - Trying connect to: xxx
UI Initialized
Webserver started
Widgets: 141
WiFi not connected. Trying connect to: xxx
WiFi reconnected
IP address: 
192.168.178.197

... => try again 5 times

MQTT not connected, reconnect...
failed, rc=-4, retrying
MQTT not connected, reconnect...
failed, rc=-4, retrying
MQTT not connected, reconnect...
failed, rc=-4, retrying
MQTT not connected, reconnect...
failed, rc=-4, retrying
MQTT not connected, reconnect...
failed, rc=-4, retrying
MQTT connection not possible, rebooting...WS_EVT_DISCONNECT

... => reboot esp after 5 attempts


! ! ! ! ! ! ! ! ! ! ! ! ! ! ! ! ! ! ! ! ! !

> WiFi Mode: AccessPoint <
1. connect your device to SSID: ESP-Buderus-km271
2. open Browser and go to Address: http://192.168.4.1

! ! ! ! ! ! ! ! ! ! ! ! ! ! ! ! ! ! ! ! ! !

... => esp has detected a "double reset" and activated the setup-mode
```

**Reply by JieffeEl** [reply]
Just tried it on an improvised test rig and it displays the same behaviour indeed.

If the double reset problem can't be solved, you could add an option in setup to "No reboot if MQTT connection lost". Or increase the delay between two reboots ?

After all, I guess a lot of us are using a Raspberry Pi or equivalent as home automation server, and they are not very fast at rebooting after an update, thus causing the unavailability of the MQTT server for too long. 

As long as the web interface still works and displays a "No MQTT server available", it should be OK.

**Reply by dewenni** [reply]
@kalleick @JieffeEl
it should be fixed now with release [v3.2.1](https://github.com/dewenni/ESP_Buderus_KM271/releases/tag/v3.2.1)
